### PR TITLE
Add Nix flake for dev shell

### PR DESCRIPTION
**Pull request by aecsocket**
NixOS ~~requires~~ strongly encourages you to have a flake.nix file defining a dev shell that you use for developing code. This PR adds such a Nix flake for a Rust dev environment. Works with rust-analyzer in VS Code as well.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,5 @@
 /target
 /chrome
 /chromedriver
+/.envrc
+/.direnv
```
Comment (MOZGIII) [inline .gitignore:5]: Why ignore those two?

**Comment by MOZGIII** [comment]
If we were to have nix stuff is this repo, I'd like to have CI flows to ensure it works and doesn't drift. Would you like to add it to this PR?

File: .gitignore (hunk 1 above)
Comment (aecsocket) [reply .gitignore:5]: .envrc is a direnv file which automatically starts a dev shell when you open the repo's directory.

.direnv is .envrc's cache, with cached flake inputs and artifacts and stuff

**Comment by aecsocket** [comment]
For the CI, I'm not sure myself how to define a CI pipeline to run Nix builds. Would this CI be alongside the existing pipeline? Would the existing pipeline be migrated to use Nix? I'm really not sure how this should work.

Considering VSC extensions and settings, I just use [direnv](https://marketplace.visualstudio.com/items?itemName=Rubymaniac.vscode-direnv) to manage it for me, so I don't have to define anything extra in the flake file.

File: .gitignore (hunk 1 above)
Comment (MOZGIII) [reply .gitignore:5]: Let's include the `.envrc` in the repo as well

**Comment by MOZGIII** [comment]
> For the CI, I'm not sure myself how to define a CI pipeline to run Nix builds. Would this CI be alongside the existing pipeline? Would the existing pipeline be migrated to use Nix? I'm really not sure how this should work.

I am intending to support both non-nix and nix workflows, so this means that nix workflow can not replace the non-nix tests. Thus, it has to be an addition rather than replacement.
In essence, we'd at the very least have to add code format checking for nix. I might add the code CI via nix later on.

File: .gitignore (hunk 1 above)
Comment (aecsocket) [reply .gitignore]: Added :+1:

**Comment by aecsocket** [comment]
I can add a formatting CI step. Only thing is which formatter to use, since Nix doesn't actually have an official one yet. I'll try some out.

**Comment by MOZGIII** [comment]
> I can add a formatting CI step. Only thing is which formatter to use, since Nix doesn't actually have an official one yet. I'll try some out.

`nil`? That's the one I'm using, but if you do your own research please share what you find as I'd like to know how the available options compare.

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    };
+  };
+  outputs = { self, nixpkgs, flake-utils, rust-overlay }:
+    flake-utils.lib.eachDefaultSystem
+      (system:
+        let
+          overlays = [ (import rust-overlay) ];
+          pkgs = import nixpkgs {
+            inherit system overlays;
+          };
```
Comment (MOZGIII) [inline flake.nix:20]: Not a fan of `import`s here, let me think about what we could do here...

Comment (MOZGIII) [reply flake.nix:20]: The `pkgs` one is good, but the `import rust-overlay` might actually be doing extra work.

Let's change it:
```suggestion
          overlays = [ rust-overlay.overlays.default ];
          pkgs = import nixpkgs {
            inherit system overlays;
          };
```

Comment (MOZGIII) [reply flake.nix:20]: Actually, nvm, since the `rust-overlay` is not accessed anywhere as a flake it works efficiently.

**Comment by MOZGIII** [comment]
Sorry, I am using not `nil` - but the `nixpkgs-fmt`

File: flake.lock
Comment (MOZGIII) [inline flake.lock:1]: This lockfile doesn't look new, let's recreate it from scratch using latest versions of the sources

**Comment by MOZGIII** [comment]
I have managed to set up my local nix system to enable more-or-less trivial development for basic rust, which this codebase is.
For this reason, I no longer see the value to maintain a nix flake here. In general, it would make sense to have a flake for something that actually produces shippable artifacts that need capturing dependencies - and this codebase is a library that handles that via crates.io. The extra weight added my the need to maintain parallel development environment definitions with nix does not really worth it.

**Pull request closed without merging**